edgedb-client/src/builder.rs

```diff
@@ -322,6 +329,171 @@ impl fmt::Display for DisplayAddr<'_> {
     }
 }
 
+impl DSN {
+    fn new(dsn: &str) -> Result<Self, Error> {
+        let admin = dsn.starts_with("edgedbadmin://");
+        if !dsn.starts_with("edgedb://") && !admin {
+            return Err(ClientError::with_message(format!(
+                "String {:?} is not a valid DSN",
+                dsn,
+            )));
+        };
+        let url = url::Url::parse(dsn).map_err(|e| {
+            ClientError::with_source(e).context(format!("cannot parse DSN {:?}", dsn))
+        })?;
+        let query = url
+            .query_pairs()
+            .into_owned()
+            .into_iter()
+            .collect::<HashMap<_, _>>();
+        Ok(Self { url, admin, query })
+    }
+
+    async fn get_value<T>(
+        &mut self,
+        key: &'static str,
+        mut v: Option<T>,
+        conv: impl FnOnce(String) -> Result<T, Error>,
+        default: impl FnOnce() -> T,
+    ) -> Result<T, Error> {
```

Comment:
Calls to this function are hard to read. Four args is not too many usually. But two lambdas and an Option and sometimes `T` is also an option, is hard to read. And also difference between `v` and `default` is a bit subtle.

I think that it would be easier if part of that is done in `retrieve_xx` (or rather `get_x`). Even if some logic is duplicated (which is `if Some(...)` and `.unwrap_or` I think).  I also think this is easier to do if conflict logic is separate from parsing logic.